### PR TITLE
Initial analytics REST APIs

### PR DESCRIPTION
**Pull request by anuruddhal**
### Proposed changes in this pull request

- Add initial analytics REST APIs
- REST time zone conversion to UTC and related changes

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/AnalyzerImpl.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    private String username;
+    private AnalyticsDAO analyticsDAO;
+
+    public AnalyzerImpl(String username, AnalyticsDAO analyticsDAO) {
+        this.username = username;
+        this.analyticsDAO = analyticsDAO;
+    }
+
+    /**
+     * @see Analyzer#getApplicationCount(Instant, Instant, String)
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/AnalyzerImpl.java]: Shall we add some debug logs to all the methods. at least to log the requested  properties and out put

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.analytics/src/main/java/org/wso2/carbon/apimgt/rest/api/analytics/impl/ApiApiServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    public Response apiListGet(String startTime, String endTime, String createdBy, Request request) throws
+            NotFoundException {
+        String username = RestApiUtil.getLoggedInUsername(request);
+        startTime = RestApiUtil.fromISO8601ToUTC(startTime);
+        endTime = RestApiUtil.fromISO8601ToUTC(endTime);
+        try {
+            if (log.isDebugEnabled()) {
+                log.debug("Retrieving API information. [From: " + startTime + " to: " + endTime + " created by: " +
+                        "" + createdBy + "]");
+            }
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.analytics/src/main/java/org/wso2/carbon/apimgt/rest/api/analytics/impl/ApiApiServiceImpl.java]: shall we use one line debug logs using lambdas as in https://garygregory.wordpress.com/2015/09/16/a-gentle-introduction-to-the-log4j-api-and-lambda-basics/

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.commons/src/main/java/org/wso2/carbon/apimgt/rest/api/common/util/RestApiUtil.java [hunk 3]
```diff
@@ -439,4 +479,33 @@ public static String getHost(String protocol) {
...
         }
     }
+
+    /**
+     * Converts time in millis to ISO-8601 date time format
+     *
+     * @param epochMilliSeconds time in milli seconds
+     * @return time in ISO 8601 format
+     */
+    public static String epochToISO8601DateTime(long epochMilliSeconds) {
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.commons/src/main/java/org/wso2/carbon/apimgt/rest/api/common/util/RestApiUtil.java]: I think these two methods should move to core

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.commons/src/main/java/org/wso2/carbon/apimgt/rest/api/common/util/RestApiUtil.java [hunk 4]
```diff
@@ -39,40 +40,47 @@
...
      */
     public static String getLoggedInUsername(Request request) {
         String loggedInUser = null;
         try {
             loggedInUser = request.getProperty(LOGGED_IN_USER).toString();
-       } catch (NullPointerException e) {
-           String message = "Error while getting loggedInUser from request";
-           log.error(message, e);
-       }
+        } catch (NullPointerException e) {
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.commons/src/main/java/org/wso2/carbon/apimgt/rest/api/common/util/RestApiUtil.java]: i know u hven't done that but we shouldnt catch null pointers

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.analytics/src/gen/java/org/wso2/carbon/apimgt/rest/api/analytics/JacksonJsonProvider.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
+package org.wso2.carbon.apimgt.rest.api.analytics;
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.analytics/src/gen/java/org/wso2/carbon/apimgt/rest/api/analytics/JacksonJsonProvider.java]: licence?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.analytics/src/gen/java/org/wso2/carbon/apimgt/rest/api/analytics/StringUtil.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
+package org.wso2.carbon.apimgt.rest.api.analytics;
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.analytics/src/gen/java/org/wso2/carbon/apimgt/rest/api/analytics/StringUtil.java]: licence?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIManagerFactory.java [hunk 7]
```diff
@@ -180,6 +195,38 @@ public APIPublisher getAPIProvider(String username) throws APIManagementExceptio
     }
 
     /**
+     * Get Analyzer for a particular user
+     *
+     * @param username The username of user who's requesting the object
+     * @return APIPublisher object
+     * @throws APIManagementException if error occurred while initializing Analytics object
+     */
+    public Analyzer getAnalyzer(String username) throws APIManagementException {
```
[the remark this answers is not in the document]
Comment (anuruddhal) [reply components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIManagerFactory.java]: All the places we have used the username.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.analytics/src/gen/java/org/wso2/carbon/apimgt/rest/api/analytics/StringUtil.java (hunk 6 above)
Comment (anuruddhal) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.analytics/src/gen/java/org/wso2/carbon/apimgt/rest/api/analytics/StringUtil.java]: This is an auto-generated file.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.analytics/src/gen/java/org/wso2/carbon/apimgt/rest/api/analytics/JacksonJsonProvider.java (hunk 5 above)
Comment (anuruddhal) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.analytics/src/gen/java/org/wso2/carbon/apimgt/rest/api/analytics/JacksonJsonProvider.java]: This is an auto-generated file.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.commons/src/main/java/org/wso2/carbon/apimgt/rest/api/common/util/RestApiUtil.java (hunk 3 above)
Comment (anuruddhal) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.commons/src/main/java/org/wso2/carbon/apimgt/rest/api/common/util/RestApiUtil.java]: This method is used only in REST APIs.

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIManagerFactory.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (pubudu538) [reply components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIManagerFactory.java]: This concept came from C4. Now we don't need to maintain user wise instances. Let's remove username. We can do the same for publisher  and store as well.

[the remark this answers is not in the document]
Comment (anuruddhal) [reply components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIManagerFactory.java]: +1.  Will send a separate PR for that.

**Pull request merged**